refactor: Use our own implementation of urlDecode

fjahr opened the pull request:
Fixes #29654 (as a side-effect)

Removing dependencies is a general goal of the project and the xz backdoor has been an additional wake up call recently. Libevent shows many of the same symptoms, few maintainers and slow releases. While libevent can not be removed completely over night we should start removing it’s usage where it's possible, ideally with the end goal to removing it completely.

This is a pretty easy win in that direction. The [`evhttp_uridecode` function from libevent](https://github.com/libevent/libevent/blob/e0a4574ba2cbcdb64bb2b593e72be7f7f4010746/http.c#L3542) we were using in `urlDecode` could be easily emulated in fewer LOC. This also ports the [applicable test vectors over from libevent](https://github.com/libevent/libevent/blob/master/test/regress_http.c#L3430).

laanwj commented:
Concept ACK
yes please... the hooking we do for `URL_DECODE` and `UrlDecodeFn` made me kind of scared, to be honest (in that context), i think can we get rid of that at the same time?

This will also fix @luke-jr's #29654.

TheCharlatan commented:
Concept ACK

hebasto commented:
Concept ACK.

stickies-v commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,46 @@
+    std::string result{"\0\0x\0\0", 5};
+    std::string encoded{"%00%00x%00%00"};
+    BOOST_CHECK_EQUAL(urlDecode(encoded).size(), 5);
+    BOOST_CHECK(urlDecode(encoded) == result);
```
nit
```suggestion
    BOOST_CHECK_EQUAL(urlDecode(encoded), result);
```

fjahr replied:
done

stickies-v commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
+#include <charconv>
 
-std::string urlDecode(const std::string &urlEncoded) {
+std::string urlDecode(const std::string& urlEncoded) {
```
We didn't use [plus decoding](https://github.com/libevent/libevent/blob/e0a4574ba2cbcdb64bb2b593e72be7f7f4010746/http.c#L3505) with `evhttp_uridecode`, so this implementation keeps functionality the same. Since `urlDecode` is a generic utility function, I think there should be a docstring that highlights that this implementation doesn't decode `+` to ` ` in case it's used for other purposes in the future.

fjahr replied:
Added a short docstring.

stickies-v commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
+                // Next two characters are part of the percent encoding
+                i += 2;
+            } else {
+                // In case of invalid percent encoding, add the '%' and continue
```
What's the rationale for continuing here? Should we accept invalid URLs? At first sight, I think I'd prefer having this return a `util::Result` and have the function fail in case of missing hex digits or invalid hex?

fjahr replied:
I also found this a bit strange but it's the behavior of libevent and they have an explicit test case for this. I didn't investigate this further yet and I think the safest option is to keep this a pure refactor for now.

laanwj replied:
For better or worse, this silent accepting behavior for invalid hex values in URL decoding is really common. See for example Python:

```python3
>>> import urllib.parse
>>> urllib.parse.unquote('%12%%xx%0a')
'\x12%%xx\n'
```
It's the web philopsophy of "in case of doubt, just do something!" 😅 . Anyhow, in this refactor it doesn't make sense imo to add error handling here.

stickies-v reviewed (commented):
Approach ACK

Since `urlDecode` no longer needs libevent, we can clean up the workaround introduced in https://github.com/bitcoin/bitcoin/pull/18504 too _(edit, oh that's basically what @laanwj said a few comments up):_

<details>
<summary>git diff on 2447bc7774</summary>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 129deeec60..830171f63a 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -11,7 +11,6 @@
 #include <clientversion.h>
 #include <common/args.h>
 #include <common/system.h>
-#include <common/url.h>
 #include <compat/compat.h>
 #include <compat/stdin.h>
 #include <policy/feerate.h>
@@ -51,7 +50,6 @@
 using CliClock = std::chrono::system_clock;
 
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
-UrlDecodeFn* const URL_DECODE = urlDecode;
 
 static const char DEFAULT_RPCCONNECT[] = "127.0.0.1";
 static const int DEFAULT_HTTP_CLIENT_TIMEOUT=900;
diff --git a/src/bitcoin-wallet.cpp b/src/bitcoin-wallet.cpp
index d5dfbbec27..fe90958a5f 100644
--- a/src/bitcoin-wallet.cpp
+++ b/src/bitcoin-wallet.cpp
@@ -11,7 +11,6 @@
 #include <clientversion.h>
 #include <common/args.h>
 #include <common/system.h>
-#include <common/url.h>
 #include <compat/compat.h>
 #include <interfaces/init.h>
 #include <key.h>
@@ -28,7 +27,6 @@
 #include <tuple>
 
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
-UrlDecodeFn* const URL_DECODE = nullptr;
 
 static void SetupWalletToolArgs(ArgsManager& argsman)
 {
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index 4f0a816388..54796c5abb 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -12,7 +12,6 @@
 #include <common/args.h>
 #include <common/init.h>
 #include <common/system.h>
-#include <common/url.h>
 #include <compat/compat.h>
 #include <init.h>
 #include <interfaces/chain.h>
@@ -35,7 +34,6 @@
 using node::NodeContext;
 
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
-UrlDecodeFn* const URL_DECODE = urlDecode;
 
 #if HAVE_DECL_FORK
 
diff --git a/src/common/url.h b/src/common/url.h
index b16b8241af..0283f471bc 100644
--- a/src/common/url.h
+++ b/src/common/url.h
@@ -7,8 +7,6 @@
 
 #include <string>
 
-using UrlDecodeFn = std::string(const std::string& url_encoded);
-UrlDecodeFn urlDecode;
-extern UrlDecodeFn* const URL_DECODE;
+std::string urlDecode(const std::string& urlEncoded);
 
 #endif // BITCOIN_COMMON_URL_H
diff --git a/src/qt/main.cpp b/src/qt/main.cpp
index ded057dbfa..16befd99e8 100644
--- a/src/qt/main.cpp
+++ b/src/qt/main.cpp
@@ -4,7 +4,6 @@
 
 #include <qt/bitcoin.h>
 
-#include <common/url.h>
 #include <compat/compat.h>
 #include <util/translation.h>
 
@@ -17,7 +16,6 @@
 extern const std::function<std::string(const char*)> G_TRANSLATION_FUN = [](const char* psz) {
     return QCoreApplication::translate("bitcoin-core", psz).toStdString();
 };
-UrlDecodeFn* const URL_DECODE = urlDecode;
 
 const std::function<std::string()> G_TEST_GET_FULL_NAME{};
 
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 2c18184261..38350b33cc 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -14,7 +14,6 @@
 #include <banman.h>
 #include <chainparams.h>
 #include <common/system.h>
-#include <common/url.h>
 #include <consensus/consensus.h>
 #include <consensus/params.h>
 #include <consensus/validation.h>
@@ -81,7 +80,6 @@ using node::RegenerateCommitments;
 using node::VerifyLoadedChainstate;
 
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
-UrlDecodeFn* const URL_DECODE = nullptr;
 
 /** Random context to get unique temp data dirs. Separate from g_insecure_rand_ctx, which can be seeded from a const env var */
 static FastRandomContext g_insecure_rand_ctx_temp_path;
diff --git a/src/wallet/rpc/util.cpp b/src/wallet/rpc/util.cpp
index 06ec7db1bc..4e93b0eac5 100644
--- a/src/wallet/rpc/util.cpp
+++ b/src/wallet/rpc/util.cpp
@@ -61,9 +61,9 @@ bool ParseIncludeWatchonly(const UniValue& include_watchonly, const CWallet& wal
 
 bool GetWalletNameFromJSONRPCRequest(const JSONRPCRequest& request, std::string& wallet_name)
 {
-    if (URL_DECODE && request.URI.substr(0, WALLET_ENDPOINT_BASE.size()) == WALLET_ENDPOINT_BASE) {
+    if (request.URI.substr(0, WALLET_ENDPOINT_BASE.size()) == WALLET_ENDPOINT_BASE) {
         // wallet endpoint was used
-        wallet_name = URL_DECODE(request.URI.substr(WALLET_ENDPOINT_BASE.size()));
+        wallet_name = urlDecode(request.URI.substr(WALLET_ENDPOINT_BASE.size()));
         return true;
     }
     return false;

```
</details>

nit: Would then update the fn signature to `std::string UrlDecode(const std::string& url_encoded)`

fjahr commented:
Addressed @laanwj 's comment and removed the hooking.

FWIW, I wondered for a moment why this is in common and not util but that was a conscious decision, see #26294.

> yes please... the hooking we do for `URL_DECODE` and `UrlDecodeFn` made me kind of scared, to be honest (in that context), i think can we get rid of that at the same time?

Absolutely, I hadn't even seen that yet. For reviewers that wonder as well, the point of this was to not have the wallet depend on `libevent`, see #18504.

> This will also fix @luke-jr's #29654.

Nice, I also missed that one!

fjahr commented:
Addressed @stickies-v 's feedback as well, thanks a lot!

laanwj commented on src/Makefile.test.include:
```diff
@@ -85,6 +85,7 @@ BITCOIN_TESTS =\
   test/checkqueue_tests.cpp \
   test/coins_tests.cpp \
   test/coinstatsindex_tests.cpp \
+	test/common_url_tests.cpp \
```
Indentation isn't right here

Sjors commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < urlEncoded.size()) {
+            int decoded_value = 0;
+            auto [ptr, ec] = std::from_chars(&urlEncoded[i + 1], &urlEncoded[i + 3], decoded_value, 16);
```
caa2040edc46291b3f032c64018eda2f4b31df39 nit: `auto [_, ec]`

fjahr replied:
done

Sjors commented on src/bitcoin-cli.cpp:
```diff
@@ -51,7 +50,6 @@
 using CliClock = std::chrono::system_clock;
 
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
-UrlDecodeFn* const URL_DECODE = urlDecode;
```
Was this doing anything, or did we forget to remove it earlier?

fjahr replied:
Not sure I understand your question, do you mean this line in particular or the whole hooking thing? This line was added the with the rest of hooking code in #18504 and I don't anything was changed about that until now, so I don't know where we might have forgotten to remove it.

Sjors reviewed (commented):
ACK 68bf6c82fee917756b367db4e99a8f6bcb6476bf

Link to the code as of v2.2.12 which we use in depends:

* `evhttp_uridecode`: https://github.com/libevent/libevent/blob/release-2.1.12-stable/http.c#L3225-L3246
* `evhttp_decode_uri_internal`: https://github.com/libevent/libevent/blob/release-2.1.12-stable/http.c#L3169-L3205

I was also happy to see that `test/fuzz/string.cpp` already covers this function.

paplorinc commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < url_encoded.size()) {
+            int decoded_value = 0;
+            auto [ptr, ec] = std::from_chars(&url_encoded[i + 1], &url_encoded[i + 3], decoded_value, 16);
```
`ptr` seems unused, we might as well inline it to:
```C++
std::from_chars(&url_encoded[i + 1], &url_encoded[i + 3], decoded_value, 16).ec == std::errc{})
```
(note, I've used `std::errc{}` instead, seems we're using that more often in the codebase)

fjahr replied:
I think this is less readable so I will keep it as is but I did take @Sjors 's suggestion to change it to change `ptr` to `_`.

paplorinc commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
+                i += 2;
+            } else {
+                // In case of invalid percent encoding, add the '%' and continue
+                res += c;
```
we could unify these two `res += c` cases by merging the two conditions:
```C++
std::string UrlDecode(const std::string& url_encoded) {
    std::string res;
    res.reserve(url_encoded.size());

    for (size_t i = 0; i < url_encoded.size(); ++i) {
        int decoded_value = 0;
        if (url_encoded[i] == '%' &&
            i + 2 < url_encoded.size() &&
            std::from_chars(&url_encoded[i + 1], &url_encoded[i + 3], decoded_value, 16).ec == std::errc{}) {
            res += static_cast<char>(decoded_value);
            i += 2;
        } else {
            res += url_encoded[i];
        }
    }

    return res;
}
```

fjahr replied:
I think this is less readable so I will keep it as is but I did take @Sjors 's suggestion to change it to change `ptr` to `_`.

fjahr replied:
Like above, I think this is a bit less readable so I will keep it as is.

paplorinc commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,46 @@
+    BOOST_CHECK_EQUAL(UrlDecode("%f0%a0%b0"), "\xf0\xa0\xb0");
+}
+
+BOOST_AUTO_TEST_CASE(decode_internal_nuls_test) {
```
is this passing for the old implementation?
Isn't this a change in behavior?

fjahr replied:
You're right, good catch! I was focussing too much on the behavior inside of libevent and didn't take into account that the result was cast to `std::string` afterwards, which lead to the behavior that a null character terminated the string. I have changed this to keep behavior consistent with the libevent version.

paplorinc replied:
Thanks for adding me as a co-author, if you force push again, please change the email to `Lőrinc <pap.lorinc@gmail.com>`

fjahr replied:
> Thanks for adding me as a co-author, if you force push again, please change the email to `Lőrinc <pap.lorinc@gmail.com>`

fixed

paplorinc commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,46 @@
+BOOST_AUTO_TEST_SUITE(common_url_tests)
+
+// These test vectors were ported from test/regress.c in the libevent library
+// which used to be a dependency of the UrlDecode function.
```
nit: it used to be a dependency of `urlDecode`, not `UrlDecode`, but I guess it's fine

fjahr replied:
Either way, someone might get confused, but I think it's better to keep it as is.

fjahr committed:
refactor: Replace libevent use in urlDecode with our own code

fjahr committed:
test: Add unit tests for urlDecode

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

fjahr committed:
refactor: Remove hooking code for urlDecode

The point of this was to be able to build bitcoin-tx and bitcoin-wallet without libevent, see #18504.

Now that we use our own implementation of urlDecode this is not needed anymore.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

fjahr committed:
scripted-diff: Modernize name of urlDecode function and param

-BEGIN VERIFY SCRIPT-
sed -i 's/urlDecode/UrlDecode/g' $(git grep -l 'urlDecode' ./src)
sed -i 's/urlEncoded/url_encoded/g' $(git grep -l 'urlEncoded' ./src)
-END VERIFY SCRIPT-

fjahr commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
+        char c = url_encoded[i];
+        // Special handling for percent which should be followed by two hex digits
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < url_encoded.size()) {
```
Thanks, added.

fjahr commented:
Addressed feedback by @paplorinc and @Sjors , thanks a lot!

I also reorganized the commits a bit because they were doing more than one thing and it was hard to address feedback cleanly. The tests are now split from the implementation, which makes it easier to cherry-pick the unit test commit and run it against master. And I split off the renaming into a scripted diff.

fjahr commented:
CI failure can be ignored, it was introduced by #29827 and should hopefully be fixed by #29926

luke-jr commented:
Concept NACK. Embedding copies of shared code just makes things more fragile, not safer. It's far more likely there's a bugfix that we miss picking up on, than that there's malicious code inserted later.

laanwj commented:
>  It's far more likely there's a bugfix that we miss picking up on, than that there's malicious code inserted later.

Sometimes i'd agree with that reasoning. But let's face it, `urlDecode` is not more complex than say, `leftpad`, it's unlikely to need maintainance or bug fixes, and thus seems unnecessary to carry a dependency for. 

(Also this isn't a direct copy of the C code! It's a C++ reimplementation that does `std::string` to `std::string`, so should handle embedded NULL bytes without truncation)

fjahr commented:
> Embedding copies of shared code just makes things more fragile, not safer. It's far more likely there's a bugfix that we miss picking up on, than that there's malicious code inserted later.

In addition to what @laanwj  said, it's also a much simpler re-implementation. The libevent function provided further options that we didn't care about and hardcoded, that's how we could simplify. And as was discussed in the last IRC meeting, there is also interest in removing libevent in general, at which point this would be a necessary change anyway. We are also removing that hooking code which is a big improvement in itself. If you subtract the test code, we are almost removing as much code as we are adding.

> It's a C++ reimplementation that does std::string to std::string, so should handle embedded NULL bytes without truncation

So, to be clear, your opinion is for this part I should [remove the patch](https://github.com/bitcoin/bitcoin/pull/29904#discussion_r1573347809) and introduce a change of behavior, right?

laanwj commented:
> So, to be clear, your opinion is for this part I should https://github.com/bitcoin/bitcoin/pull/29904#discussion_r1573347809 and introduce a change of behavior, right?

Yes. There's no use in emulating broken C string behavior.

fjahr committed:
common: Don't terminate on null character in UrlDecode

The previous behavior was the result of casting the result returned from the libevent function evhttp_uridecode to std:string but this was probably not intended.

fjahr commented:
I have added the behavior change in a separate commit.

paplorinc commented on src/common/url.h:
```diff
@@ -7,8 +7,10 @@
-extern UrlDecodeFn* const URL_DECODE;
+/* Decode a URL.
+ *
+ * Notably this implementation does not decode a '+' to a ' '.
```
would it make sense to mention the `&` as well?

fjahr replied:
Are these commonly decoded as a space as well? I couldn't find evidence for this but I didn't spend much time on it.

paplorinc replied:
Not exactly: https://github.com/libevent/libevent/blob/release-2.1.12-stable/http.c#L3186-L3188

fjahr replied:
Hm, the code you are linking to is part of how libevent handles the `+`. Or did you want to refer to `?`? I think if it's not very common for `&` to be decoded as space then we don't need to mention it.

paplorinc replied:
? seems to be a control char, `If -1, when true we transform plus to space only after we've seen a ?.  -1 is deprecated.`, no need to mention it, thanks for checking.

laanwj reviewed (approved):
Concept and code review ACK 175d57ba953cdc1da3a54ae0208dad813f7ea4cf

hebasto reviewed (commented):
Is `if USE_LIBEVENT` guard in https://github.com/bitcoin/bitcoin/blob/175d57ba953cdc1da3a54ae0208dad813f7ea4cf/src/Makefile.am#L715-L718 still needed?

laanwj commented:
No, it isn't. Nor is piling on the `EVENT_CFLAGS`. Good catch @hebasto.

fjahr commented:
> No, it isn't. Nor is piling on the `EVENT_CFLAGS`.

Removed both, thanks @hebasto and @laanwj !

hebasto commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
nit: For new source files we can use a copyright header with HTTPS link and future-proof dating:
```suggestion
// Copyright (c) 2024-present The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or https://opensource.org/license/mit/.
```

hebasto commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,65 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/util/setup_common.h>
```
nit: It seems no symbols from that header are used in the source file.

fjahr replied:
done

hebasto reviewed (commented):
Approach ACK b7a3f8e375b87e4f3ccfa7994905dfbf46a23bf2.

stickies-v reviewed (approved):
ACK b7a3f8e375b87e4f3ccfa7994905dfbf46a23bf2

Good to remove this unnecessary libevent dependency, and nice work on restructuring the commits etc. Easy to review PR. Left some nits and happy to quickly review if you decide to address them.

---

Not essential for this PR, but I noticed we're doing two unnecessary string copies (through `std::string::substr()`) in code touched by this PR:

<details>
<summary>git diff on b7a3f8e375</summary>

```diff
diff --git a/src/common/url.cpp b/src/common/url.cpp
index cd2c38dc44..fddbd6ab15 100644
--- a/src/common/url.cpp
+++ b/src/common/url.cpp
@@ -5,10 +5,11 @@
 #include <common/url.h>
 
 #include <string>
+#include <string_view>
 #include <stdexcept>
 #include <charconv>
 
-std::string UrlDecode(const std::string& url_encoded) {
+std::string UrlDecode(std::string_view url_encoded) {
     std::string res;
     res.reserve(url_encoded.size());
 
diff --git a/src/common/url.h b/src/common/url.h
index c1df626142..203f41c70f 100644
--- a/src/common/url.h
+++ b/src/common/url.h
@@ -6,11 +6,12 @@
 #define BITCOIN_COMMON_URL_H
 
 #include <string>
+#include <string_view>
 
 /* Decode a URL.
  *
  * Notably this implementation does not decode a '+' to a ' '.
  */
-std::string UrlDecode(const std::string& url_encoded);
+std::string UrlDecode(std::string_view url_encoded);
 
 #endif // BITCOIN_COMMON_URL_H
diff --git a/src/wallet/rpc/util.cpp b/src/wallet/rpc/util.cpp
index 2e7b1449bc..f1fb9db3d2 100644
--- a/src/wallet/rpc/util.cpp
+++ b/src/wallet/rpc/util.cpp
@@ -14,6 +14,7 @@
 #include <univalue.h>
 
 #include <boost/date_time/posix_time/posix_time.hpp>
+#include <string_view>
 
 namespace wallet {
 static const std::string WALLET_ENDPOINT_BASE = "/wallet/";
@@ -61,9 +62,9 @@ bool ParseIncludeWatchonly(const UniValue& include_watchonly, const CWallet& wal
 
 bool GetWalletNameFromJSONRPCRequest(const JSONRPCRequest& request, std::string& wallet_name)
 {
-    if (request.URI.substr(0, WALLET_ENDPOINT_BASE.size()) == WALLET_ENDPOINT_BASE) {
+    if (request.URI.starts_with(WALLET_ENDPOINT_BASE)) {
         // wallet endpoint was used
-        wallet_name = UrlDecode(request.URI.substr(WALLET_ENDPOINT_BASE.size()));
+        wallet_name = UrlDecode(std::string_view{request.URI}.substr(WALLET_ENDPOINT_BASE.size()));
         return true;
     }
     return false;

```
</details>

fjahr commented on src/Makefile.am:
```diff
@@ -712,10 +712,7 @@ libbitcoin_common_a_SOURCES = \
 
-if USE_LIBEVENT
-libbitcoin_common_a_CPPFLAGS += $(EVENT_CFLAGS)
 libbitcoin_common_a_SOURCES += common/url.cpp
```
done

fjahr commented on src/common/url.cpp:
```diff
@@ -4,19 +4,33 @@
-
-#include <cstdlib>
 #include <string>
+#include <stdexcept>
```
both done

fjahr commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,65 @@
+    BOOST_CHECK_EQUAL(UrlDecode("%f0%a0%b0"), "\xf0\xa0\xb0");
+}
+
+BOOST_AUTO_TEST_CASE(decode_internal_nuls_test) {
```
done, I took this naming from libevent exactly fwiw but since we change the behavior and tests anyway, it's not that relevant anymore to keep it consistent

fjahr commented:
Addressed latest feedback from @hebasto and @stickies-v , thank you!

hebasto commented on src/Makefile.am:
```diff
@@ -679,6 +679,7 @@ libbitcoin_common_a_SOURCES = \
   common/run_command.cpp \
   common/settings.cpp \
   common/system.cpp \
+	common/url.cpp \
```
Indentation?

fjahr replied:
should be fixed now, weird, I must have some plugin that inserts these tabs :/

fjahr commented on src/common/url.cpp:
```diff
@@ -4,19 +4,34 @@
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < url_encoded.size()) {
+            int decoded_value{0};
+            auto [_, ec] = std::from_chars(&url_encoded[i + 1], &url_encoded[i + 3], decoded_value, 16);
```
Yeah, I just pushed this.

fjahr commented on src/common/url.cpp:
```diff
@@ -4,19 +4,34 @@
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < url_encoded.size()) {
+            int decoded_value{0};
+            auto [_, ec] = std::from_chars(url_encoded.data() + i + 1, url_encoded.data() + i + 1, decoded_value, 16);
```
fixed, thanks

stickies-v reviewed (approved):
ACK 477c03b42825084ac344050d97ea206a82ba0eb6

Apologies for my UB string_view code suggestion, I didn't realize `operator[pos]` behaves differently for string_view vs string when `pos==size()`. What you pushed looks like an elegant fix while keeping the copying minimal, nice.

laanwj commented on configure.ac:
```diff
@@ -1706,7 +1706,6 @@ AM_CONDITIONAL([ENABLE_QT_TESTS], [test "$BUILD_TEST_QT" = "yes"])
 AM_CONDITIONAL([ENABLE_BENCH], [test "$use_bench" = "yes"])
 AM_CONDITIONAL([USE_QRCODE], [test "$use_qr" = "yes"])
 AM_CONDITIONAL([USE_LCOV], [test "$use_lcov" = "yes"])
-AM_CONDITIONAL([USE_LIBEVENT], [test "$use_libevent" = "yes"])
```
Even this can go? It's fascinating to see how this little function was such a wedge issue and now we can drop half the build system 😄

laanwj reviewed (approved):
Code review ACK 477c03b42825084ac344050d97ea206a82ba0eb6

fjahr commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,64 @@
+BOOST_AUTO_TEST_CASE(decode_malformed_test) {
+    BOOST_CHECK_EQUAL(UrlDecode("%%xhello th+ere \xff"), "%%xhello th+ere \xff");
+
+    BOOST_CHECK_EQUAL(UrlDecode("%"), "%");
```
Hm, ok, that's another behavior change I didn't anticipate. Let me first describe my understanding of happens: In [the libevent version](https://github.com/libevent/libevent/blob/e0a4574ba2cbcdb64bb2b593e72be7f7f4010746/http.c#L3492) each of the two characters following the `%` are [looked at individually if they are valid hex characters](https://github.com/libevent/libevent/blob/e0a4574ba2cbcdb64bb2b593e72be7f7f4010746/http.c#L3508). The space is not so the if fails and the fallback case takes over, which just adds the `%` to res and then the next two chars are processed individually and since they are not special cases by themselves, they are also just added to res, so the result is `%1 `. In the current version here we look at both characters together with `std::from_chars`. The white space seems to be ignored, no error raised, and so the `decoded_value` is integer `1`. Then the `static_cast<char>(decoded_value)` converts the `1` to a non-printable character and so the result visually appears to be an empty string. It seems to me now that it's safer to check both digits individually if they are hex, so I have changed the code in this regard and added a few more tests with spaces. Let me know if you found other cases where the behavior is still changing.

paplorinc replied:
There's still one sneaky difference that I've found: `%00`.
Previously this prematurely terminated the string, now it's kept:
```C++
BOOST_CHECK_EQUAL(urlDecode("a%00b"), std::string("a\u0000b", 3));
```
this failed previously, passes now.

paplorinc replied:
Yes https://github.com/bitcoin/bitcoin/pull/29904#discussion_r1572090699

paplorinc replied:
Those cases were added since, I see, thanks.

maflcko reviewed (approved):
lgtm, but `unsigned` may be better

fjahr commented on src/common/url.cpp:
```diff
@@ -4,19 +4,42 @@
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < urlEncoded.size()) {
+            int decoded_value{0};
+            auto [p, ec] = std::from_chars(urlEncoded.data() + i + 1, urlEncoded.data() + i + 3, decoded_value, 16);
```
neat, I applied this change

fjahr replied:
Ugh, looks like MSVC has a problem with this, I am reverting it for now (see https://github.com/bitcoin/bitcoin/actions/runs/8823024109/job/24222527982?pr=29904)

fanquake replied:
Another MSVC bug for us to report upstream I guess. @hebasto.

fjahr commented on src/common/url.cpp:
```diff
@@ -4,19 +4,42 @@
+        // Special handling for percent which should be followed by two hex digits
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < urlEncoded.size()) {
+            int decoded_value{0};
```
Changed and added a test for this case as well

fjahr commented:
> `unsigned` may be better

No idea what this means ;) but I addressed the feedback, thanks a lot!

theStack reviewed (approved):
Code-review ACK 992c714451676cee33d3dff49f36329423270c1c

maflcko reviewed (approved):
Did some differential fuzzing, as this is an ideal fit for it.

ACK 992c714451676cee33d3dff49f36329423270c1c 👈

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 992c714451676cee33d3dff49f36329423270c1c 👈
uoUAes8qa1RR+FiK2t8YFFjOnn0fR1ewcF3JZcxlbFYRLH099vwrEWYq5uCUNGpAtidxMgd/deCwvPSTZbvgDQ==
```

</details>

l0rinc reviewed (commented):
~I think we should document at least that we thought of the `%00` corner case, e.g. via a test case or a change.~ (already added)

I have verified that the behavior is otherwise the same before and after via:
```C++
std::string GenerateRandomString() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> length_dis(1, 10);
    std::uniform_int_distribution<> char_dis(1, 255);

    size_t length = length_dis(gen);
    std::string res;
    res.reserve(length);

    for (size_t i = 0; i < length; ++i) {
        char c = static_cast<char>(char_dis(gen));
        res.push_back(c);
    }

    return res;
}

std::string urlDecode_old(const std::string &urlEncoded) {
    std::string res;
    if (!urlEncoded.empty()) {
        char *decoded = evhttp_uridecode(urlEncoded.c_str(), false, nullptr);
        if (decoded) {
            res = std::string(decoded);
            free(decoded);
        }
    }
    return res;
}

BOOST_AUTO_TEST_CASE(randomized_url_decode_length_test) {
    for (int i = 0; i < 100'000'000; ++i) {
        std::string random_encoded = GenerateRandomString();

        std::string result1 = urlDecode_old(random_encoded);
        std::string result2 = urlDecode(random_encoded);

        if (result1 != result2) {
            std::cerr
                << "urlDecode discrepancy found, random_encoded = `" + random_encoded + "`, urlDecode_old = `" +
                   result1 + "`, urlDecode (new) = `" + result2 + "`" << std::endl;
        }

        BOOST_CHECK_LE(result1.size(), random_encoded.size());
    }
}
```

stickies-v reviewed (approved):
ACK 992c714451676cee33d3dff49f36329423270c1c

nit: definitely not worth more back-and-forth, but I think (untested) maflcko's [suggestion](https://github.com/bitcoin/bitcoin/pull/29904#discussion_r1578145369) should work with MSVC when explicitly converting iterators to pointers:

<details>
<summary>git diff on 992c714451</summary>

```diff
diff --git a/src/common/url.cpp b/src/common/url.cpp
index ecf88d07ea..c901ecb5dc 100644
--- a/src/common/url.cpp
+++ b/src/common/url.cpp
@@ -14,25 +14,24 @@ std::string UrlDecode(std::string_view url_encoded)
     std::string res;
     res.reserve(url_encoded.size());
 
-    for (size_t i = 0; i < url_encoded.size(); ++i) {
-        char c = url_encoded[i];
+    for (auto it = url_encoded.begin(); it < url_encoded.end(); ++it) {
         // Special handling for percent which should be followed by two hex digits
         // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
-        if (c == '%' && i + 2 < url_encoded.size()) {
+        if (*it == '%' && std::distance(it, url_encoded.end()) > 2) {
             unsigned int decoded_value{0};
-            auto [p, ec] = std::from_chars(url_encoded.data() + i + 1, url_encoded.data() + i + 3, decoded_value, 16);
+            auto [p, ec] = std::from_chars(&*(it + 1), &*(it + 3), decoded_value, 16);
 
             // Only if there is no error and the pointer is set to the end of
             // the string, we can be sure both characters were valid hex
-            if (ec == std::errc{} && p == url_encoded.data() + i + 3) {
+            if (ec == std::errc{} && p == &*(it + 3)) {
                 res += static_cast<char>(decoded_value);
                 // Next two characters are part of the percent encoding
-                i += 2;
+                it += 2;
                 continue;
             }
             // In case of invalid percent encoding, add the '%' and continue
         }
-        res += c;
+        res += *it;
     }
 
     return res;

```
</details>

fjahr commented on src/test/common_url_tests.cpp:
```diff
@@ -0,0 +1,72 @@
+    BOOST_CHECK_EQUAL(UrlDecode("%20invalid%ZX"), " invalid%ZX");
+    BOOST_CHECK_EQUAL(UrlDecode("%20%G1%ZX"), " %G1%ZX");
+
+    BOOST_CHECK_EQUAL(UrlDecode("%1 "), "%1 ");
```
thanks, I will add them when I have to retouch, if not I will make a small follow-up

fjahr commented:
> I think (untested) maflcko's [suggestion](https://github.com/bitcoin/bitcoin/pull/29904#discussion_r1578145369) should work with MSVC when explicitly converting iterators to pointers

Yes, it's certainly possible and I briefly explored it myself, but it didn't seem like much of an improvement over the current version anymore. And the current version had already received some review at that point, so I decided it would be better to just revert since it was just a style-nit anyway.

achow101 commented:
ACK cb8f83df1916308013a6ae4b8e4914eb805c5343

ryanofsky commented:
I went ahead and merged this, despite NACK from luke https://github.com/bitcoin/bitcoin/pull/29904#issuecomment-2067871490 about costs of forking shared code. It seems like the NACK was adequately responded to, with responses acknowledging the tradeoff and giving reasons why it seemed justified in this case (even fixing a bug luke reported https://github.com/bitcoin/bitcoin/issues/29654)

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/181.

pinheadmz commented on src/common/url.cpp:
```diff
@@ -4,19 +4,36 @@
+                i += 2;
+                continue;
+            }
+            // In case of invalid percent encoding, add the '%' and continue
```
I'm curious about the discrepancy between this and https://github.com/bitcoin/bitcoin/pull/27468. They relate to libevent decode uri and parse uri methods respectively and the latter is more sensitive to invalid `%` sequences (ie that aren't exactly two hexadecimal characters). 

I believe it's an example of lenient acceptance from users but strict sending to other machines. However in both of our use cases (wallet init args, and the REST interface) strict RFC3689 adherence doesn't seem necessary. 

As I'm reimplementing HTTP without libevent it seems like this is better way to go, and relax the test added by #27468.

Thoughts ?

pablomartin4btc replied:
Having chatted with you in #31194 I do agree with your thoughts.

I've tested this PR adding these 2 checks (comparing it with the libevent uri decoder [output](https://github.com/bitcoin/bitcoin/issues/31194#issuecomment-2585885945)), similar to what you did in your [branch](https://github.com/pinheadmz/bitcoin/blob/4084f6f739412acea0df7c93a2d31bdeeadef9bb/src/test/httpserver_tests.cpp#L110-L111) (sorry to be late on this review):

```
    BOOST_CHECK_EQUAL(UrlDecode("%3Fp1=v1%20&p1=v2%20"), "?p1=v1 &p1=v2 ");
    BOOST_CHECK_EQUAL(UrlDecode("%3Fp1%3Dv1%26p2%3D100%25"), "?p1=v1&p2=100%");
```

In #27468 httprequest/ REST (even before that change) uses libevent uri parser (without calling the libevent uri decoder, which this PR replaced the only instance we had of it with `UrlDecode()`), and next you will replace the libevent uri parser by using this `UrlDecode()` (your current [branch](https://github.com/bitcoin/bitcoin/compare/master...pinheadmz:bitcoin:http-rewrite)).

Guy790 commented:
Duplicate of #
